### ISSUE TITLE
New Check: WhenInSwitchStatement

### ISSUE THREAD

**Issue by Devansh-18155** [checkstyle/checkstyle#21364]
### Describe the problem in detail:

Java 21 introduced pattern matching for switch statements and expressions along with when guard clauses (JEP 441). Currently, `WhenShouldBeUsedCheck` is restricted to switch rules `(case ... ->) `to maintain the existing switch syntax form.

There is currently no check that inspects traditional switch statement groups `(case ...:)` to report cases where a single if statement inside a pattern case block could be simplified using a when guard clause.

This new check will inspect traditional switch statement groups with pattern labels and suggest using when guard clauses, while safely skipping case groups that involve(#15199).

**Input Code (Violation):**
```java
public class Test {
    public void process(Object obj) {
        switch (obj) {
            case String s:
                if (s.length() > 5) {
                    System.out.println(s);
                }
                break;
            default:
                break;
        }
    }
}
```
**Expected refactored code:**
```java
public class Test {
    public void process(Object obj) {
        switch (obj) {
            case String s when s.length() > 5:
                System.out.println(s);
                break;
            default:
                break;
        }
    }
}
```
```bash
/var/tmp $ cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
    <module name="WhenInSwitchStatement"/>
  </module>
</module>
/var/tmp $ javac Test.java
/var/tmp $ cat Test.java
public class Test {
    public void process(Object obj) {
        switch (obj) {
            case String s:
                if (s.length() > 5) {
                    System.out.println(s);
                }
                break;
            default:
                break;
        }
    }
}
/var/tmp $ RUN_LOCALE="-Duser.language=en -Duser.country=US"
/var/tmp $ java $RUN_LOCALE -jar checkstyle-14.0.0-all.jar -c config.xml Test.java
Starting audit...
[ERROR] /var/tmp/Test.java:5:17: Using 'when' is preferred over a single 'if' statement inside a case block. [WhenInSwitchStatement]
Audit done.
```
### Describe how the new check will solve the problem. Mention the required properties and their behaviour:

The proposed `WhenInSwitchStatement` check will:

- Target `CASE_GROUP AST` tokens in traditional switch statements.

- Verify if a pattern case label contains a single if statement without an else branch.

- Ensure the case group ends with a control flow statement (break, yield, return, throw) to avoid flagging cases with intentional  (#15199).

**Comment by romani** [comment]
Please strictly follow issue template

**Comment by Devansh-18155** [comment]
@romani  , I have updated the issue with the template. Can you take a look at it.

**Comment by ZaheerAhmadDev** [comment]
@Devansh-18155 
For ` : ` please be aware of fallthrough there is a case where it silently bypasses a when guard.
Please see the example in: https://github.com/checkstyle/checkstyle/pull/15199

**Comment by romani** [comment]
We need new Check for this.
As goal of WhenShouldBeUsedCheck to not change form of switch.
What you try to do is convertion of simply old style switch to new style.

**Comment by Devansh-18155** [comment]
`WhenShouldBeUsedCheck` is meant to keep the existing switch style rather than changing old switch statements to new ones and (#15199) needs careful handling.
Since a new check is needed for this so should I update this issue's title and description for the `New Check` proposal or should I open a new issue for it

**Comment by ZaheerAhmadDev** [comment]
@Devansh-18155 please update issue title and description for new check to lets us see new check design.please follow this template of new check  https://github.com/checkstyle/checkstyle/blob/master/docs/report-issue.md#how-to-request-new-checkmodule

**Comment by Devansh-18155** [comment]
@romani  @ZaheerAhmadDev , I have made a new Check for this issue can you review once and am on it if approved!

**Comment by romani** [comment]
Please share expected format of switch. In issue description example of CLI 

Sounds like your Check is about conversion of old style switch to new style if possible, am I right ?

**Comment by Devansh-18155** [comment]
@romani No, it does not convert old-style switch (`:`) to new-style arrow rules (`->`).
It keeps the traditional colon switch style (`case String s when s.length() > 5:`) and just replaces the inner `if` block with a `when` guard clause.
I've updated the issue description to show the expected refactored code format as well!

**Comment by abhinav-phi** [comment]
Hi! I'd like to take this new check.
Plan: I'll implement the `WhenInSwitchStatement` check per the spec in this issue, follow the existing Check patterns, and add unit tests + xdoc documentation.
Could you please assign this to me? Thanks!

/assign take

**Comment by romani** [comment]
@abhinav-phi , for new contributors we recommend
Please [read](https://github.com/checkstyle/checkstyle/blob/master/.github/CONTRIBUTING.md#getting-started) and watch playlist videos at [Starting_Development](https://github.com/checkstyle/checkstyle/blob/master/docs/BEGINNING_DEVELOPMENT.md#starting-development).
